File: 01-presentation-blog-examples/src/main/java/com/nalepa/demo/example04/Example04_Many_Blocking_IO.java
package com.nalepa.demo.example04;

import java.time.LocalTime;

public class Example04_Many_Blocking_IO {
    public static void main(String[] args) {
        simulateBlockingIO(1);
        simulateBlockingIO(2);
        simulateBlockingIO(3);
    }

    static void simulateBlockingIO(int index) {
        System.out.println();
        log("Start blocking IO code for: " + index);
        nonThrowingSleep(5000);
        log("End blocking IO code for: " + index);
    }

    static void log(String message) {
        System.out.println(Thread.currentThread().getName() + " : " + LocalTime.now() + " : " + message);
    }

    static void nonThrowingSleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }
}

File: presentation-blog-examples/src/main/kotlin/com/nalepa/demo/example04/Example04_Many_Blocking_IO.kt
package com.nalepa.demo.example04

import java.time.LocalTime

fun main() {
    simulateBlockingIO(1)
    simulateBlockingIO(2)
    simulateBlockingIO(3)
}

fun simulateBlockingIO(index: Int) {
    println("")
    log("Start blocking IO code for: $index")
    Thread.sleep(5000) // http call, database call, reading from file, whatever what needs external data
    log("End blocking IO code for: $index")
}

fun log(message: String) {
    println("${Thread.currentThread().name} : ${LocalTime.now()} : $message")
}

